design/design-doc-package-refactor.md

```diff
@@ -0,0 +1,480 @@
+     with the name of the `Package`.
+    - Note: If there are other `PackageRevision`s owned by this `Package` that
+      have state `Pending` they should be marked as `Inactive`. This could occur
+      if a previous upgrade was deemed unsafe or a `Package` was updated
+      multiple times in quick succession.
+4. Compare the `Pending` `PackageRevision` to the `Active` `PackageRevision` and
+   determine if it is safe to upgrade. If it is not safe to upgrade, do not
+   proceed. User will be required to change `spec.package` back to previous
+   version if they want to cancel upgrade. Until doing so, the previous
+   `PackageRevision` will continue to be `Active`.
+  - Safe upgrade will initially only be a check that support for CRDs for which
+    instances currently exist or are depended on by other `Package`s will not be
+    dropped, but can be expanded to be more sophisticated in the future.
+5. If upgrade is deemed safe, the `Active` `PackageRevision` should be marked as
+   `Stopping`, and controller should wait until the controller deployed by that
+   `PackageRevision` has been stopped (i.e. `Deployment` no longer exists).
+6. When the `Stopping` `PackageRevision` has finished cleaning up its
+   controller, it should be marked as `Inactive` and the new `PackageRevision`
+   should be marked as `Active`. This will cause it to install all of its
+   objects (updating any existing ones), and start its own controller (i.e.
+   create a `Deployment`).
+7. Wait for `PackageRevision` to become `Ready` (this will occur after
+   successful install of all objects included in the `PackageRevision`).
+8. Set the `status.currentHash` of the `Package` to match the `sha256` hash of
+   the package image and `status.currentImage` to match the package image name.
+
+
+*Deleting a `Package`*
+
+1. Observe delete of a `Package`.
+2. Get the `Active` `PackageRevision` and check that there are no other `Active`
+   package revisions (for other `Package`s) that depend on CRDs that are managed
+   by this `Package` and mark as `Stopping`.
+3. When the `Stopping` `PackageRevision` has finished cleaning up its controller
+   it should be marked as `Inactive`. and the _`Package` controller_ should
+   clean up any CRDs, `InfrastructureDefinition`s, `InfrastructurePublication`s,
+   and `Composition`s that have a label corresponding to the `Package`.
+4. All `PackageRevision`s that are labeled with `Package` name should be
+   deleted.
+
+*Optional Features*
+
+The following are optional features that could be implemented in a future
+iterations:
+
+- Block deletion until all CRDs are manually cleaned up.
+- Only retain a specified number of `Inactive` `PackageRevision`s. This could be
+  accomplished by adding a field to the `Package` `spec` such as
+  `revisionHistoryLimit`. The _`Package` controller_ would then clean up any
+  `PackageRevision`s that are not within the `revisionHistoryLimit`.
+
+#### PackageRevision Controller
```

Comment:
I am less suspicious about whether we need a new type `PackageRevision` after reading this section but honestly I couldn't imaging how it looks like before seeing the actual YAML. Also, it could be helpful explaining that we had `*PackageInstall -> Package` coupling but now `Package -> N PackageRevisions` and why this is better. I can see it's operationally safer to have multiple revisions as snapshots for upgrades but I can't immediately see how an actual rollback would work.

Reply:
I am going back and forth on the need for `PackageRevision`. The major value proposition is that if you fail to upgrade, then want to roll back and the image of the previous has changed (i.e. same tag but different hash), then you have a snapshot of the configuration as it was. That being said, we are not keeping around the install `Job`, so we don't actually have a full snapshot. I originally had the `PackageRevision` controller creating the install `Job` and then switched it, but I think I may need to go back to that model and have the install `Job` for a revision existing as long as the `PackageRevision` itself does.

Reply:
> The major value proposition is that if you fail to upgrade, then want to roll back and the image of the previous has changed (i.e. same tag but different hash), then you have a snapshot of the configuration as it was.

Actually, I'm still suspicious about the feasibility of a package roll-back. Let's say I updated provider-gcp:v0.10 to v0.11 and many CRDs got updated, new ones added and some of them are removed. Between the updates, there have been CRs created/updated. In case of rollback, are we able to reverse these operations without breaking the custom resources without breaking anything?

I said that I'm less suspicious about its need thinking `PackageRevision` object could make the manual operations safer but I can't immediately see how it helps if it doesn't make roll-back possible.

Reply:
> Let's say I updated provider-gcp:v0.10 to v0.11 and many CRDs got updated, new ones added and some of them are removed.

@muvaf the idea is that if an upgrade is not safe, then we don't actually try to modify the CRDs. So the rollback would be employed in a case where an update was deemed unsafe and so we want to roll back to activating the old revision (i.e. starting its controller). That being said, you could theoretically have a situation where you upgraded successfully and still wanted to roll back, in which case it would be just like an upgrade but an "older" version. Let's say for instance that you were running `provider-gcp:v0.10.0` and you updated to `provider-gcp:v0.10.1` which had only a change in controller behavior. You then realized that the new controller behavior was not actually something you wanted, so you rolled back to `provider-gcp:v0.10.0`. This should be possible if there is no change in CRD support.

Reply:
> the idea is that if an upgrade is not safe, then we don't actually try to modify the CRDs. So the rollback would be employed in a case where an update was deemed unsafe and so we want to roll back to activating the old revision (i.e. starting its controller). 

Let me understand what you're thinking. I have updated the version and there are some checks to validate everything and until all validations pass we don't take any critical action like CRD update or stopping controller. If checks fail, we show an error in the status saying we can't do that. Does the revision come into play at that point? Instead of having user to switch to the earlier version, we have them switch to the earlier revision number?

Reply:
@muvaf yep it is mostly just convenience, but could also help with troubleshooting if they are able to compare the `spec` of multiple versions of the same package.

Reply:
@hasheddan To be honest, I don't think the convenience it adds justifies the complexity that `PackageRevision` type adds. In fact, changing the version back instead of revision number sounds even easier to me.